node/network/statement-distribution/src/lib.rs

```diff
@@ -246,6 +246,7 @@ impl<R: rand::Rng> StatementDistributionSubsystem<R> {
 				},
 			};
 
+			// TODO: Call this regularly instead of after every message?
 			vstaging::dispatch_requests(&mut ctx, &mut state).await;
```

Comment:
> I am thinking now that calling dispatch_requests after every message might not be the best way. In addition to not being called when a request becomes ready to retry, it also seems unnecessary, as not all handled messages may lead to pending requests. In fact, only a fraction actually result in pending requests.

Yeah, I remember seeing this code & concluding that it was fine until we added timeouts when reviewing. Polling periodically is ~ok but also less than ideal. It may be better to rewrite the main loop with some control flow like this:

```
fn dispatch_requests() {
  // dispatch all requests that are ready. wake up on every new timeout and try again. loop forever
}

fn main() {
    loop {
        match select(dispatch_requests(), receive_muxed()) {
            // ...
        }
    }
}
```

This architecture would interrupt the request-dispatching logic whenever we got a new message, handling it, and then dispatching request and waiting for new messages in the background.

I think ordering of `dispatch_requests()` and `receive_muxed()` matters because we don't want to delay sending out requests if a lot of messages are coming in.

Reply:
Ah, I had guessed at your intention behind `dispatch_requests` and figured you would call it out in review if I got it wrong. I should have checked with you though.

So are you saying we should keep the existing `dispatch_requests`, but just change it to loop? What do you have in mind for "wake up on every new timeout and try again"? Are you thinking there would be some sleep while looping, which would be preempted when a request is ready to retry? When would we recalculate the closures?

Since order matters, do you have in mind the [`select_biased!`](https://docs.rs/futures/latest/futures/macro.select_biased.html) macro? I don't see how that would work since `dispatch_requests` would now be looping forever and would never yield, hence would never be selected. I think we would have to make the function yield requests which are then dispatched in the body of the select. Whereas if `receive_muxed` is selected, then we would handle that, loop back around and call `dispatch_requests` again, and re-calculate the closures, which I think is what we want.

Reply:
The order in which they're polled does matter, though. If `receive_muxed_message()` resolves on the first `poll()` before `dispatch_requests` gets polled, then the result will be that we receive messages without dispatching requests.

Reply:
I'm not attached to any architecture, just keep in mind that we would prefer not to poll on a timer. If you don't like this solution then please try another one.

Reply:
> The order in which they're polled does matter, though.

Yes, agreed. I was just trying to understand your proposal and make sure it works. I think I wasn't clear -- I agree we should use select (specifically `select_biased!` as regular `select!` would poll randomly instead of prioritizing `dispatch_requests`). But if `dispatch_requests` is in a forever loop and never yielding then that wouldn't work. I'm trying to make sure I'm not missing something though, because parts of your proposal e.g. "wake up on every new timeout and try again" weren't clear to me.

All that said though, I think I do see a clear path forward -- not sure if it's what you had in mind, but if you're agnostic to the final architecture, I'll go ahead with something that should work.

Reply:
As discussed, I started by implementing simple timeout-based preemption. But it seemed like an ugly hack and I felt bad, so I changed it to use your soonest-retry idea. Hope it looks good.